source/common/unified_memory_allocation/include/uma/memory_provider.h

```diff
@@ -12,6 +12,31 @@
 #include <uma/base.h>
 #include <uma/memory_provider_ops.h>
 
+#define UMA_ATTR_NORMAL 0          /* No further special treatment.  */
+#define UMA_ATTR_RANDOM 1          /* Expect random page references.  */
+#define UMA_ATTR_SEQUENTIAL 2      /* Expect sequential page references.  */
+#define UMA_ATTR_WILLNEED 3        /* Will need these pages.  */
+#define UMA_ATTR_DONTNEED 4        /* Don't need these pages.  */
+#define UMA_ATTR_FREE 8            /* Free pages only if memory pressure.  */
+#define UMA_ATTR_REMOVE 9          /* Remove these pages and resources.  */
+#define UMA_ATTR_DONTFORK 10       /* Do not inherit across fork.  */
+#define UMA_ATTR_DOFORK 11         /* Do inherit across fork.  */
+#define UMA_ATTR_MERGEABLE 12      /* KSM may merge identical pages.  */
+#define UMA_ATTR_UNMERGEABLE 13    /* KSM may not merge identical pages.  */
+#define UMA_ATTR_HUGEPAGE 14       /* Worth backing with hugepages.  */
+#define UMA_ATTR_NOHUGEPAGE 15     /* Not worth backing with hugepages.  */
+#define UMA_ATTR_DONTDUMP 16       /* Explicity exclude from the core dump, */
+                                   /* overrides the coredump filter bits.  */
+#define UMA_ATTR_DODUMP 17         /* Clear the UMA_ATTR_DONTDUMP flag.  */
+#define UMA_ATTR_WIPEONFORK 18     /* Zero memory on fork, child only.  */
+#define UMA_ATTR_KEEPONFORK 19     /* Undo UMA_ATTR_WIPEONFORK.  */
+#define UMA_ATTR_COLD 20           /* Deactivate these pages.  */
```

Comment:
What deactivate means?

Reply:
From https://man7.org/linux/man-pages/man2/madvise.2.html :
```
       MADV_COLD (since Linux 5.4)
              Deactivate a given range of pages.  This will make the
              pages a more probable reclaim target should there be a
              memory pressure.  This is a nondestructive operation.  The
              advice might be ignored for some pages in the range when
              it is not applicable.
```
All attributes are copied from https://sourceware.org/git/?p=glibc.git;a=blob;f=sysdeps/unix/sysv/linux/bits/mman-linux.h;h=6771f07a0f9a617fcde0eb264d54784788d22e76;hb=HEAD#l86.
We can narrow down the number of them of course.

Reply:
We need only `MADV_DONTNEED, MADV_HUGEPAGE, MADV_NOHUGEPAGE` in qkmalloc.

Reply:
For now, I would define just the few things that you need.

Reply:
For MADV_DONTNEED, I think it would be better to implement something like 'purge' in jemalloc. We will need that anyway if we want to implement the 'decay' functionally we discussed some time ago. I'm not sure if there is a good way to implement `purge` for USM/device memory (or we can just call something like zeVirtualMemFree?), but it's probably fine.

As for MADV_HUGEPAGE and MADV_NOHUGEPAGE: I'm not convinced if we should expose those at all. For example, there is no equivalent to MADV_NOHUGEPAGE on Windows, and requesting HUGE_PAGES can only be done during virtual memory reservation, not after. It might be better to just break the abstraction here and on linux, just call `madvise` directly instead of passing it through set_attrs.